### ISSUE TITLE
Use PMPI to detect application code calling MPI under VT scheduler

### ISSUE THREAD

**Issue by PhilMiller** [DARMA-tasking/vt#510]
**Is your feature request related to a problem? Please describe.**

In EMPIRE, we found on a few occasions bugs in adapting code for VT execution where something down the stack was calling MPI routines where we didn't expect. If VT had a way of detecting that this was happening and alerting on it, debugging them would have moved much faster. If other MPI codes are to be adapted for VT, we'll want tooling to facilitate this.

**What Needs to be Done?**

I've adapted some PMPI code generation from https://github.com/UIUC-PPL/PMPI_Projections to emit wrappers for the list of MPI routines, check whether a flag saying VT is running is set, and call `vtAbort` indicating the offending call if so.

To bake this into VT, we would want to internally suppress warnings about our own usage, and otherwise alert on any MPI call below the scheduler on the stack.

**Describe potential solution outcome**

**Describe alternatives you've considered**

I considered trying to script this in `gdb`, but figured that it was going to be very finicky to write even for a single process, and extremely hard to run reliably in parallel on arbitrary target machines.

**Additional context**

**Comment by lifflander** [comment]
@PhilMiller I think this would be a great feature to add to VT. Where is the code you adapted for this? Does this work by using `LD_PRELOAD`?

**Comment by PhilMiller** [comment]
The generated code is on a branch in EMPIRE-PIC right now, 310-mpi-..., since it was initially a one-off for ad-hoc dynamic analysis of a bug we were concerned about.

I did this by adapting the code generation and function list from https://github.com/UIUC-PPL/PMPI_Projections which is licensed as Illinois/NCSA (BSD-equivalent). Would a derivative of that be acceptable for incorporation here?

It can be made to work via `LD_PRELOAD`, though I baked it directly into the binary in EMPIRE for the expedience of not having to think about linkage in the initial use.

What I'd do with it in VT is conditionally compile the wrappers in debug-mode builds, and omit them in release-mode builds. Debug would see a minor overhead for code calling MPI, and release would see none, with nothing having to be done at runtime, and no need for special effort by the user at compile or launch time to get the benefits.
